libs/features/commerce/src/placeholder.js

```diff
@@ -0,0 +1,153 @@
+import Service from './service.js';
+
+export const FAILED = 'placeholder-failed';
+export const PENDING = 'placeholder-pending';
+export const RESOLVED = 'placeholder-resolved';
+
+const setImmediate = (callback) => setTimeout(callback, 0);
+
+/**
+ * Stores private bucket of values for each instance of placeholder.
+ * @type {WeakMap<HTMLPlaceholderMethods, {
+ *  error?: Error;
+ *  promises: {
+ *    resolve: (HTMLPlaceholderMethods) => void;
+ *    reject: (error: Error) => void;
+ *  }[];
+ *  state: FAILED | PENDING | RESOLVED | undefined;
+ *  timer?: NodeJS.Timeout;
+ *  version: number;
+ * }>}
+ */
+const buckets = new WeakMap();
+
+/** @type {Commerce.HTMLPlaceholderMixin & Record<string, any>} */
+// @ts-ignore
+export const HTMLPlaceholderMethods = {
+  attributeChangedCallback(name, _, value) {
+    this.log?.debug(`Attribute "${name}":`, value);
+    this.update();
+  },
+
+  connectedCallback() {
+    this.log?.debug('Connected:', this.parentElement);
+    this.update();
+  },
+
+  init() {
+    buckets.set(this, {
+      promises: [],
+      state: undefined,
+      version: 0,
+    });
+  },
+
+  onceSettled() {
```

Comment:
I would avoid using a new term. We call placeholders resolved on success, so `onceResolved` sounds more natural to me.

Reply:
@yesil 

The `resolved` term has strong association with `Promise.resolve` static method.
But former `onceResolved` method returns promises that can be also `rejected` when associated placeholder fails. This can be confusing for consumers because `rejected` is an opposite to `resolved`.
The new `onceSettled` name is proposed for analogy with [Promise.allSettled](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Promise/allSettled) static method.
For backward compatibility, the old `onceResolved` name is also kept.